presto-main/src/test/java/com/facebook/presto/spiller/BenchmarkAESSpillCipher.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spiller;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+
+@State(Scope.Benchmark)
+@Fork(1)
+@OutputTimeUnit(TimeUnit.SECONDS)
+@Warmup(batchSize = 1, iterations = 1, time = 10, timeUnit = TimeUnit.SECONDS)
+@Measurement(batchSize = 1, iterations = 1, time = 3, timeUnit = TimeUnit.SECONDS)
+@BenchmarkMode(Mode.Throughput)
+public class BenchmarkAESSpillCipher
```

Comment:
- This benchmark is missing main method. How do you run it? 
- @Param("4196") is unnecessary because there is only one value
- Usually, we create a BenchmarkData class with setup method and add BenchmarkData parameter to benchmark methods like below
- It is helpful to add a @Test method to do some basic validation - it will run on each commit and ensure that benchmark doesn't get terribly broken over time.

```
@State(Scope.Thread)
@Fork(2)
@Warmup(iterations = 10, time = 500, timeUnit = TimeUnit.MILLISECONDS)
@Measurement(iterations = 10, time = 500, timeUnit = TimeUnit.MILLISECONDS)
@OutputTimeUnit(TimeUnit.MILLISECONDS)
@BenchmarkMode(Mode.Throughput)
public class BenchmarkAESSpillCipher
{
    @State(Scope.Thread)
    public static class BenchmarkData
    {
        private final AesSpillCipher spillCipher = new AesSpillCipher();
        private byte[] plainText;
        private byte[] cipherText;

        @Setup
        public void setup()
                throws IOException
        {
            this.plainText = createRandomBytes(4196);
            this.cipherText = spillCipher.encrypt(plainText);
        }
    }

    @Benchmark
    public byte[] decryptArray(BenchmarkData data)
    {
        return data.spillCipher.decrypt(data.cipherText);
    }

    @Benchmark
    public byte[] encryptArray(BenchmarkData data)
    {
        return data.spillCipher.encrypt(data.plainText);
    }

    @Benchmark
    public ByteBuffer decryptBuffer(BenchmarkData data)
    {
        return data.spillCipher.decrypt(ByteBuffer.wrap(data.cipherText));
    }

    @Benchmark
    public ByteBuffer encryptBuffer(BenchmarkData data)
    {
        return data.spillCipher.encrypt(ByteBuffer.wrap(data.plainText));
    }

    private static byte[] createRandomBytes(int size)
    {
        byte[] bytes = new byte[size];
        ThreadLocalRandom.current().nextBytes(bytes);
        return bytes;
    }

    @Test
    public void verify()
            throws IOException
    {
        BenchmarkData data = new BenchmarkData();
        data.setup();
        BenchmarkAESSpillCipher benchmark = new BenchmarkAESSpillCipher();
        benchmark.decryptArray(data);
        benchmark.decryptBuffer(data);
        benchmark.encryptArray(data);
        benchmark.encryptBuffer(data);
    }

    public static void main(String[] args)
            throws RunnerException
    {
        Options options = new OptionsBuilder()
                .verbosity(VerboseMode.NORMAL)
                .include(".*" + BenchmarkAESSpillCipher.class.getSimpleName() + ".*")
                .build();
        new Runner(options).run();
    }
}
```

Here are the results from above:

```
# Run complete. Total time: 00:01:31

Benchmark                               Mode  Cnt    Score    Error   Units
BenchmarkAESSpillCipher.decryptArray   thrpt   20  124.728 ±  9.553  ops/ms
BenchmarkAESSpillCipher.decryptBuffer  thrpt   20  135.322 ± 10.958  ops/ms
BenchmarkAESSpillCipher.encryptArray   thrpt   20   85.980 ±  3.884  ops/ms
BenchmarkAESSpillCipher.encryptBuffer  thrpt   20   81.870 ±  5.894  ops/ms
```

The errors are pretty high (~10%). The benchmark setting may need to be tuned. However, it is not clear to me whether this benchmark is needed at all. If not, feel free to just delete.

I also noticed that benchmark results you posted have high error (10%). I don't think this is due to any of your changes, but it says that the benchmark needs some fixing to be useful.

Reply:
I think removing the benchmark is fine. We carried it internally because the original spill encryption PR that is superseded by this one was significantly less performant. I don't think it has value since it's the sole implementation now.